topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java

```diff
@@ -274,4 +282,95 @@ public ActorFuture<ClusterTopology> initialize() {
+
+    public RollingUpdateAwareInitializerV83ToV84(
+        final ClusterMembershipService membershipService,
+        final StaticConfiguration staticConfiguration,
+        final ConcurrencyControl executor) {
+      this.membershipService = membershipService;
+      staticInitializer = new StaticInitializer(staticConfiguration);
+      staticClusterSize = staticConfiguration.clusterMembers().size();
+      this.executor = executor;
+      initializeFuture = new CompletableActorFuture<>();
+    }
+
+    @Override
+    public ActorFuture<ClusterTopology> initialize() {
+      if (staticClusterSize == 1) {
+        // Cluster will be initialized via normal static initializer
+        initializeFuture.complete(ClusterTopology.uninitialized());
+        return initializeFuture;
+      }
+
+      final Version version = membershipService.getLocalMember().version();
+      if (isVersion8_4(version)) {
+        final boolean knowOtherMembers = hasOtherReachableMembers(membershipService);
+        if (knowOtherMembers && isRollingUpdate(membershipService)) {
+          LOGGER.debug(
+              "Cluster is doing rolling update. Cannot initialize cluster topology via gossip. Initializing cluster topology from static configuration.");
+          staticInitializer.initialize().onComplete(initializeFuture);
+        } else if (!knowOtherMembers) {
+          LOGGER.debug(
+              "No other members are reachable. Cannot initialize topology. Will retry in {}",
+              RETRY_DELAY);
+          executor.schedule(RETRY_DELAY, this::initialize);
+        } else {
+          // do not initialize. It is not rolling update so initialize via gossip or sync
+          initializeFuture.complete(ClusterTopology.uninitialized());
+        }
+      } else {
+        initializeFuture.complete(ClusterTopology.uninitialized());
+      }
+      return initializeFuture;
+    }
+
+    private boolean hasOtherReachableMembers(final ClusterMembershipService membershipService) {
+      return membershipService.getMembers().stream()
+          .map(Member::id)
+          .anyMatch(memberId -> !memberId.equals(membershipService.getLocalMember().id()));
+    }
+
+    private boolean isRollingUpdate(final ClusterMembershipService membershipService) {
+      final List<Member> otherMembers =
+          membershipService.getMembers().stream()
+              .filter(member -> !member.id().equals(membershipService.getLocalMember().id()))
+              .toList();
+      final var cannotDetermineVersionOfOtherMembers =
+          otherMembers.stream().map(Member::version).noneMatch(Objects::nonNull);
+      if (cannotDetermineVersionOfOtherMembers) {
+        // This is mainly required for testing, where the DiscoveryMembershipProvider cannot
+        // determine version of remote members.
+        LOGGER.warn(
+            "Cannot determine version of local member. Assuming this is not rolling update and skip initialization.");
```

Comment:
:x: The message is misleading, because here we cannot determine version of remote members.